homeassistant/components/drop_connect/sensor.py

```diff
@@ -0,0 +1,280 @@
+    ),
+    DROPSensorEntityDescription(
+        key=CARTRIDGE_2_LIFE,
+        translation_key=CARTRIDGE_2_LIFE,
+        icon=GAUGE_ICON,
+        native_unit_of_measurement=PERCENTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=0,
+        value_fn=lambda device: device.drop_api.cart2(),
+    ),
+    DROPSensorEntityDescription(
+        key=CARTRIDGE_3_LIFE,
+        translation_key=CARTRIDGE_3_LIFE,
+        icon=GAUGE_ICON,
+        native_unit_of_measurement=PERCENTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=0,
+        value_fn=lambda device: device.drop_api.cart3(),
+    ),
+]
+
+# Defines which sensors are used by each device type
+DEVICE_SENSORS: dict[str, list[str]] = {
+    DEV_HUB: [
+        AVERAGE_WATER_USED,
+        BATTERY,
+        CURRENT_FLOW_RATE,
+        CURRENT_SYSTEM_PRESSURE,
+        HIGH_SYSTEM_PRESSURE,
+        LOW_SYSTEM_PRESSURE,
+        PEAK_FLOW_RATE,
+        WATER_USED_TODAY,
+    ],
+    DEV_SOFTENER: [
+        BATTERY,
+        CAPACITY_REMAINING,
+        CURRENT_FLOW_RATE,
+        CURRENT_SYSTEM_PRESSURE,
+    ],
+    DEV_FILTER: [BATTERY, CURRENT_FLOW_RATE, CURRENT_SYSTEM_PRESSURE],
+    DEV_LEAK_DETECTOR: [BATTERY, TEMPERATURE],
+    DEV_PROTECTION_VALVE: [
+        BATTERY,
+        CURRENT_FLOW_RATE,
+        CURRENT_SYSTEM_PRESSURE,
+        TEMPERATURE,
+    ],
+    DEV_PUMP_CONTROLLER: [CURRENT_FLOW_RATE, CURRENT_SYSTEM_PRESSURE, TEMPERATURE],
+    DEV_RO_FILTER: [
+        CARTRIDGE_1_LIFE,
+        CARTRIDGE_2_LIFE,
+        CARTRIDGE_3_LIFE,
+        INLET_TDS,
+        OUTLET_TDS,
+    ],
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the DROP sensors from config entry."""
+    _LOGGER.debug(
+        "Set up sensor for device type %s with entry_id is %s",
+        config_entry.data[CONF_DEVICE_TYPE],
+        config_entry.entry_id,
+    )
+
+    if config_entry.data[CONF_DEVICE_TYPE] in DEVICE_SENSORS:
+        async_add_entities(
+            DROPSensor(hass.data[DOMAIN][config_entry.entry_id], sensor)
+            for sensor in SENSORS
+            if sensor.key in DEVICE_SENSORS[config_entry.data[CONF_DEVICE_TYPE]]
+        )
+
+
+class DROPSensor(DROPEntity, SensorEntity):
+    """Representation of a DROP sensor."""
+
+    entity_description: DROPSensorEntityDescription
+
+    def __init__(
+        self,
+        coordinator: DROPDeviceDataUpdateCoordinator,
+        entity_description: DROPSensorEntityDescription,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(entity_description.key, coordinator)
+        self.entity_description = entity_description
+
+    @property
+    def native_value(self) -> float | int | None:
+        """Return the value reported by the sensor."""
+        return self.entity_description.value_fn(self.coordinator)
```

Comment:
```suggestion
        if (
            value := self.entity_description.value_fn(self.coordinator)
        ) == -1 and self.state_class == SensorStateClass.MEASUREMENT:
            return None
        return value
```

Work-a-round

Reply:
The DROP firmware will now return `null' for any invalid values.  This change is reflected in the unit test data in 762b22822fc476d5497b1dbf56ee2febc5a2ae66.